bin/src/_main.rs

```diff
@@ -139,110 +142,170 @@ pub async fn _main(
-                            pod_name.unwrap_or_default(),
-                            namespace.unwrap_or_default(),
-                            pod_label.unwrap_or_default(),
-                            Arc::new(k8s_event_stream_feature_meta),
-                        ))
-                    }
-                };
-
-                let pod_name = std::env::var("POD_NAME").ok();
-                let namespace = std::env::var("NAMESPACE").ok();
 
-                let metric_stats_stream = match config.log.log_metric_server_stats {
-                    K8sTrackingConf::Never => None,
-                    K8sTrackingConf::Always => {
-                        let created_metrics_stream_feature_meta = set_up_leader(
-                            &namespace,
-                            &k8s_client,
-                            &pod_name,
-                            REPORTER_LEASE_NAME.to_string(),
-                        )
-                        .await;
+                        let (stream_handle, abort_registration) = AbortHandle::new_pair();
+                        let (thread_handle, swap_intent) = match middleware
+                            .kick_over(middleware_k8s_client.unwrap(), node_name.as_deref())
+                        {
+                            Ok((stream, swap_intent)) => {
+                                debug!("k8s metadata watcher kicked over!");
+                                let abortable_stream = Abortable::new(stream, abort_registration);
+                                (
+                                    tokio::spawn(async move {
+                                        debug!("k8s metadata watcher thread started, processing events!");
+                                        abortable_stream.for_each(|_| async {}).await;
+                                        debug!("k8s metadata watcher thread exiting!");
+                                    }),
+                                    swap_intent,
+                                )
+                            }
+                            Err(e) => {
+                                let message = format!(
+                                    "The agent could not access k8s api after several attempts: {}",
+                                    e
+                                );
+                                error!("{}", message);
+                                panic!("{}", message);
+                            }
+                        };
+
+                        handle_queue.push_back((thread_handle, stream_handle, Some(swap_intent)));
+                        if handle_queue.len() == 2 {
+                            debug!("delaying k8s metadata watcher rotation so new store can populate...");
+                            // when rotating watchers, introduce a small delay so the new
+                            // watcher has a chance to populate its cache
+                            sleep(K8S_WATCHER_ROTATION_DELAY).await;
```

Comment:
how about to use a channel here? for explicit handshake.

Reply:
outdated